performance.Rmd

```diff
@@ -0,0 +1,184 @@
+               "South Korea", "Vietnam", "South Africa", "Senegal", "Iran")
+
+set.seed(123)
+test = data.frame(
+  country = sample(countries, 1e6, TRUE),
+  x = sample(1:100, 1e6, TRUE),
+  y = sample(1:1000, 1e6, TRUE)
+)
+
+bench::mark(
+  sort_filter = {
+    tmp = test[order(test$country), ]
+    subset(tmp, country %in% c("United Kingdom", "Japan", "Vietnam"))
+  },
+  filter_sort = {
+    tmp = subset(test, country %in% c("United Kingdom", "Japan", "Vietnam"))
+    tmp[order(tmp$country), ]
+  }
+)
+```
+
+
+## How does `polars` help?
+
+We have seen that the order in which functions are applied matters a lot. But it
+already takes a long time to deal with large data, we're not going to spend even
+more time trying to optimize our pipeline, right?
+
+This is where *lazy* execution comes into play. The idea is that we write our
+code as usual, but this time, we won't apply it directly on a dataset but on a
+lazy dataset, i.e a dataset that is not loaded in memory yet (in `polars` terms,
+these are `DataFrame`s and `LazyFrame`s). Once our code is ready, we
+call `collect()` at the end of the pipeline. Before executing our code, `polars`
+will internally check whether it can be optimized, for example by reordering
+some operations.
+
+Let's re-use the example above but this time with `polars` syntax and 10M
+observations. It doesn't make sense to convert a dataset that we already have in
+memory to a `LazyFrame` so first we have to export it as CSV and then we can
+read it lazily with `lazy_csv_reader()`:
+
+```{r echo = FALSE, eval = !exists("data/test.csv")}
+set.seed(123)
+test = data.frame(
+  country = sample(countries, 1e7, TRUE),
+  x = sample(1:100, 1e7, TRUE),
+  y = sample(1:1000, 1e7, TRUE)
+)
+readr::write_csv(test, "data/test.csv")
+```
+
+
+```{r}
+library(polars)
+
+df_test = csv_reader("data/test.csv")
+lf_test = lazy_csv_reader("data/test.csv")
+```
+
+Now, we can convert the base R code above to a `polars` query:
+
+```{r}
+df_test$
+  sort(pl$col("country"))$
+  filter(
+    pl$col("country")$is_in(pl$lit(c("United Kingdom", "Japan", "Vietnam")))
+  )
+```
+
+This works for the `DataFrame`, that uses eager execution. For the `LazyFrame`, we can write the same query:
+
+```{r}
+lazy_query = lf_test$
+  sort(pl$col("country"))$
+  filter(
+    pl$col("country")$is_in(pl$lit(c("United Kingdom", "Japan", "Vietnam")))
+  )
+
+lazy_query
+```
+
+However, this doesn't do anything to the data until we call `collect()` at the 
+end. We can now compare the two approaches:
+
+```{r}
+bench::mark(
+  eager = df_test$
+    sort(pl$col("country"))$
+    filter(
+      pl$col("country")$is_in(pl$lit(c("United Kingdom", "Japan", "Vietnam")))
+    ),
+  lazy = lazy_query$collect(),
+  iterations = 10
+)
+```
+
+
+On this very simple query, using lazy execution instead of eager execution lead
+to a 1.7-2.2x decrease in time. 
+
+So what happened? Under the hood, `polars` reorganized the query so that it filters
```

Comment:
... reorganized the query **so that it filters rows already while reading the csv into memory**, and then sorts the remaining data. This can be seen by comparing ...

Reply:
done